### PR TITLE
Revert blocks during reorg

### PR DESCRIPTION
**Pull request by junderw**
## Re-index not required (but highly recommended, to remove incorrect old data from the DB)

---

Take 2

Alternative to #111

### CODE REVIEW

File: src/new_index/schema.rs [hunk 1]
```diff
@@ -273,13 +280,53 @@ impl Indexer {
...
+                    warn!("reorg of over 10 blocks detected! Wonky stuff might happen!");
+                }
+                let (tx, rx) = crossbeam_channel::unbounded();
+                // Delete txstore
+                start_fetcher(self.from, &daemon, reorged.clone())?
+                    .map(|blocks| self.add(&blocks, Operation::DeleteBlocks));
+                // Delete history_db
+                start_fetcher(self.from, &daemon, reorged)?.map(|blocks| {
+                    self.index(&blocks, Operation::DeleteBlocksWithHistory(tx.clone()))
+                });
```
Comment (mononaut) [inline src/new_index/schema.rs]: is it possible for a reorg to be detected while we're still doing the initial sync (i.e. while `self.from == FetchFrom::BlkFiles`)?

if so, does this mean the `blkfiles_fetcher` would have to go back and re-parse every block file twice here to fetch the reorged ones? for the two delete operations?

I suppose it's a fairly unlikely edge case, but I'm just imagining the initial sync getting stuck reading all of the testnet3 block files forever because the chain keeps getting reorged every 10 minutes 😅

File: src/new_index/schema.rs [hunk 2]
```diff
@@ -273,13 +280,53 @@ impl Indexer {
...
+            let mut headers = self.store.indexed_headers.write().unwrap();
+            let reorged = headers.apply(new_headers.clone());
+            assert_eq!(tip, *headers.tip());
+            let headers_len = headers.len();
+            drop(headers);
+
+            if !reorged.is_empty() {
+                if reorged.len() > 10 {
+                    warn!("reorg of over 10 blocks detected! Wonky stuff might happen!");
+                }
```
Comment (mononaut) [inline src/new_index/schema.rs]: is there anything special about 10 blocks here, or is it just a arbitrary threshold for "damn that's a big reorg"?

**Review by mononaut — COMMENTED** [review]
Concept ACK

This definitely feels like the right approach, and stepping through the code the implementation all makes sense to me.

I'll upgrade to a full ACK once I figure out how to test it properly :)

File: src/new_index/schema.rs (hunk 1 above)
Comment (mononaut) [reply src/new_index/schema.rs]: oh, I suppose that could only happen once at most each time electrs gets restarted, since immediately after this section we run the indexing all the way through the finish and then set the "initial sync completed" flag in the db 🤔

File: src/new_index/schema.rs (hunk 2 above)
Comment (junderw) [reply src/new_index/schema.rs]: arbitrary

File: src/new_index/schema.rs (hunk 1 above)
Comment (junderw) [reply src/new_index/schema.rs]: The headers get fetched once and then the initial block files sync runs for a couple days.

The headers are synced from Bitcoind before the first update call anyways, so pulling this check to the front might cause problems with initial sync if a reorg happens just between when the headers sync in the init function of the headerslist and when its called inside update...

Not sure how to test that edge case... but it should be rare enough to not matter imo.

Comment (mononaut) [reply src/new_index/schema.rs]: yeah I think it would only happen in practice if you rebooted electrs partway through the initial sync, which should never really need to happen in production anyway.

File: src/elements/asset.rs [hunk 3]
```diff
@@ -178,11 +178,17 @@ pub fn index_confirmed_tx_assets(
     network: Network,
     parent_network: BNetwork,
     rows: &mut Vec<DBRow>,
+    op: &Operation,
 ) {
     let (history, issuances) = index_tx_assets(tx, network, parent_network);
 
     rows.extend(history.into_iter().map(|(asset_id, info)| {
-        asset_history_row(&asset_id, confirmed_height, tx_position, info).into_row()
+        let history_row = asset_history_row(&asset_id, confirmed_height, tx_position, info);
```
Comment (junderw) [inline src/elements/asset.rs:186]: I think maybe we should also be sending over the issuances that are being added below this diff...?

The comment seems to indicate that the issuances DBRows are redundant (they also exist as history entries)...

File: src/new_index/schema.rs [hunk 4]
```diff
@@ -268,18 +284,68 @@ impl Indexer {
...
+            .map(|blocks| self.index(&blocks, Operation::DeleteBlocksWithHistory(tx.clone())));
+        // All senders must be dropped for receiver iterator to finish
+        drop(tx);
+
+        // All senders are dropped by now, so the receiver will iterate until the
+        // end of the unbounded queue.
+        let scripts = rx.into_iter().collect::<HashSet<_>>();
+        for script in scripts {
+            // cancel the script cache DB for these scripts. They might get incorrect data mixed in.
+            self.store.cache_db.delete(vec![
```
Comment (junderw) [inline src/new_index/schema.rs:311]: Maybe instead, we should only run delete if we can get() the keys? Not sure which would be faster.

Seems like the extra get() would add extra overhead when we delete, but be faster than attempting a delete if there is no row... averaging out to being faster overall? Would need to measure it...

Probably premature optimization concerns, just noting the concern.

File: src/elements/asset.rs (hunk 3 above)
Comment (mononaut) [reply src/elements/asset.rs:186]: yes, probably.

the `b"i"` issuance rows are used in the `lookup_asset()` method, so otherwise I think we'd leave behind phantom assets in the event that an initial issuance transaction got reorged out but not re-confirmed in a subsequent block.

Comment (mononaut) [reply src/elements/asset.rs:186]: on the other hand, can Liquid even have reorgs? 😅

File: src/new_index/schema.rs (hunk 4 above)
Comment (mononaut) [reply src/new_index/schema.rs:311]: I suspect the difference isn't worth worrying about, but we can always change it later if profiling reveals a big difference.

**Comment by junderw** [comment]
Never mind. I take back my take back! lol

The channel sendbacks are just for `b'I'` rows which are included in the cache for assets. Only sending back AssetIds from `b'I'` rows is correct, Since we want to remove them from the cache.

**Review by mononaut — CHANGES_REQUESTED** [review]
Tested NACK @ [737d7252]

I've been running this on mainnet locally for a few days, and not observed any problems at all.

To test the reorg behavior more thoroughly, I scripted the following reproducible scenarios with a pair of regtest nodes:

| # | Scenario | Master branch | comments | This PR | comments |
|---|----------|---------------|----------|---------|----------|
| 1 | Simple short reorg | ⁉️ | reorged coinbase remains in txstore, but is filtered out of address results | ✅ | reorged coinbases removed from both txstore and history dbs |
| 2 | Transaction reorged and reconfirmed in another block at the same height and index | ✅ |  | ✅ |  |
| 3 | Transaction reorged and reconfirmed in another block at the same height but a different index | ❌ | transaction is filtered out of `/address/.../txs`, but double-counted towards txos in chain_stats | ✅ | |
| 4 | Transaction reorged and reconfirmed in a later block at the same index | ✅ | | ✅ | |
| 5 | Long-range reorg (10 blocks) | ✅ |  | ❌🔥 | electrs crashes |

For short reorgs, this PR fixes the immediate issue of transactions confirmed at the same height and different indexes being double-counted.

However, long reorgs reliably crash electrs on this PR, sometimes into an irrecoverable state:

```
DEBUG - applying 11 new headers from height 103
DEBUG - reorg of 10 blocks detected
DEBUG - hash=73f96de01d42ae6349ae1c2e468b9ce6ae19cd0417f8864fc4ee87a799cbecc8 height=103 @ 2024-12-24T22:06:42Z (10 left to index)
TRACE - [THREAD] GETHASHMAP INSERT | bitcoind_fetcher-34 ThreadId(291)
TRACE - [THREAD] START WORK        | bitcoind_fetcher-34 ThreadId(291)
TRACE - [THREAD] FINISHED WORK     | bitcoind_fetcher-34 ThreadId(291)
TRACE - [THREAD] GETHASHMAP REMOVE | bitcoind_fetcher-34 ThreadId(291)
TRACE - [THREAD] HASHMAP REMOVED   | bitcoind_fetcher-34 ThreadId(291)
DEBUG - Deleting 10 blocks to Indexer
DEBUG - deleting 110 rows from RocksDB { path: "/electrs/db/regtest/newindex/txstore" }
DEBUG - hash=73f96de01d42ae6349ae1c2e468b9ce6ae19cd0417f8864fc4ee87a799cbecc8 height=103 @ 2024-12-24T22:06:42Z (10 left to index)
TRACE - [THREAD] GETHASHMAP INSERT | bitcoind_fetcher-35 ThreadId(292)
TRACE - [THREAD] START WORK        | bitcoind_fetcher-35 ThreadId(292)
TRACE - [THREAD] FINISHED WORK     | bitcoind_fetcher-35 ThreadId(292)
TRACE - [THREAD] GETHASHMAP REMOVE | bitcoind_fetcher-35 ThreadId(292)
DEBUG - Indexing (Deleting) 10 blocks with Indexer
TRACE - [THREAD] HASHMAP REMOVED   | bitcoind_fetcher-35 ThreadId(292)
thread 'thread 'lookup-txo-12lookup-txo-8' panicked at ' panicked at src/new_index/schema.rssrc/new_index/schema.rs:1448:29:
missing txo 0c674ca8df296ab312e4125be3032d9904221f0b646a89d9bab67092996b8b09:0 in DB { db: RocksDB { path: "/electrs/db/regtest/newindex/txstore" } }
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
thread 'thread 'thread 'lookup-txo-2lookup-txo-15thread ':1448:' panicked at src/new_index/schema.rs:1448:29thread 'lookup-txo-7' panicked at lookup-txo-3src/new_index/schema.rs:' panicked at 1448' panicked at 29:src/new_index/schema.rs:
missing txo 5d94c442d19918c3002bbcc641a09a84370fcb36807f69fd7b81c4d0ccd712fa:1 in DB { db: RocksDB { path: "/electrs/db/regtest/newindex/txstore" } }29lookup-txo-4thread 'thread '
```
_(note: I added a couple of extra debug logs to this trace, so line numbers etc might not line up exactly)_

the panic seems to be triggered by the reorg handling itself, rather than any collision or race condition with a simultaneous API request. My guess is that we're trying to `lookup_txos` *after* the corresponding transactions have already been deleted from the txstore, but I haven't quite teased out the exact sequence of events.

**Comment by junderw** [comment]
This should remove all issues. We will sequentially fetch 100 blocks at a time from bitcoind in reverse header order and delete them in order from the tip to the stem... so it shouldn't cause issues...

(crosses fingers)

**Review by mononaut — APPROVED** [review]
Tested ACK @ [a4549944]

Still no problems running locally on mainnet, and now it passes all of my scripted regtest reorg scenarios:

| # | Scenario | Master branch | comments | This PR | comments |
|---|----------|---------------|----------|---------|----------|
| 1 | Simple short reorg | ⁉️ | reorged coinbase remains in txstore, but is filtered out of address results | ✅ | reorged coinbases removed from both txstore and history dbs |
| 2 | Transaction reorged and reconfirmed in another block at the same height and index | ✅ |  | ✅ |  |
| 3 | Transaction reorged and reconfirmed in another block at the same height but a different index | ❌ | transaction is filtered out of `/address/.../txs`, but double-counted towards txos in chain_stats | ✅ | |
| 4 | Transaction reorged and reconfirmed in a later block at the same index | ✅ | | ✅ | |
| 5 | Long-range reorg (10 blocks) with mempool re-entry | ✅ |  | ✅ | |

**Comment by junderw** [comment]
2 minds are better than one, good catch on the flipped order.

LGTM on the changes you made.
